arwen/cryptoapi/cryptoei.go

```diff
@@ -287,3 +390,616 @@ func v1_3_verifySecp256k1(
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(data))
+	}
+
+	ec, err := managedType.GetEllipticCurve(ecHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	byteLen := (ec.BitSize + 7) / 8
+	if int(length) != 1+byteLen {
+		arwen.WithFault(arwen.ErrLengthOfBufferNotCorrect, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	xResult, yResult, err := managedType.GetTwoBigInt(xResultHandle, yResultHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoBigIntUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	managedType.ConsumeGasForBigIntCopy(ec.P, ec.N, ec.B, ec.Gx, ec.Gy, xResult, yResult)
+	xResultUC, yResultUC := elliptic.UnmarshalCompressed(ec, data)
+	if xResultUC == nil || yResultUC == nil || !ec.IsOnCurve(xResultUC, yResultUC) {
+		arwen.WithFault(arwen.ErrPointNotOnCurve, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	xResult.Set(xResultUC)
+	yResult.Set(yResultUC)
+	return 0
+}
+
+//export v1_3_generateKeyEC
+func v1_3_generateKeyEC(
+	context unsafe.Pointer,
+	xPubKeyHandle int32,
+	yPubKeyHandle int32,
+	ecHandle int32,
+	resultOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	metering := arwen.GetMeteringContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	curveMultiplier := managedType.Get100xCurveGasCostMultiplier(ecHandle)
+	if curveMultiplier < 0 {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	if curveMultiplier == 250 {
+		curveMultiplier = 500
+	}
+	gasToUse := metering.GasSchedule().CryptoAPICost.MarshalCompressedECC * uint64(curveMultiplier) / 100
+	metering.UseGas(gasToUse)
+
+	ec, err := managedType.GetEllipticCurve(ecHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoEllipticCurveUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+
+	xPubKey, yPubKey, err := managedType.GetTwoBigInt(xPubKeyHandle, yPubKeyHandle)
+	if err != nil {
+		arwen.WithFault(arwen.ErrNoBigIntUnderThisHandle, context, runtime.CryptoAPIErrorShouldFailExecution())
+		return 1
+	}
+	managedType.ConsumeGasForBigIntCopy(ec.P, ec.N, ec.B, ec.Gx, ec.Gy, xPubKey, yPubKey)
+
+	result, xPubKeyGK, yPubKeyGK, err := elliptic.GenerateKey(ec, rand.Reader)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(result))
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return int32(len(result))
+	}
+
+	xPubKey.Set(xPubKeyGK)
+	yPubKey.Set(yPubKeyGK)
+	return 0
+}
+
+//export v1_3_p224Ec
+func v1_3_p224Ec(context unsafe.Pointer) int32 {
```

Comment:
so the developer can use only a set of different elliptic curves, no ?
Wouldn't it be better to have only one function like func v1_3_createEC(context, type as string) ? Then you have a switch here according to the string constants ? It would be easier to integrate in rust as well.